### ISSUE TITLE
1.9-rc1 regression - `StaticArray` allocates `Core.SimpleVector` objects, runtime dispatch

### ISSUE THREAD

**Issue by BioTurboNick** [JuliaLang/julia#49145]
1.9-rc1:
Time:
![image](https://user-images.githubusercontent.com/1438610/227668080-6d64c780-b484-48bb-bfe1-238c18fb8c77.png)

Allocations:
![image](https://user-images.githubusercontent.com/1438610/227667931-a5a008fd-06d0-46dd-8800-f093bc93f787.png)

1.8.5:
Time:
![image](https://user-images.githubusercontent.com/1438610/227668178-e7ba20e7-27f7-4949-baee-ba4a53fff235.png)

Allocations:
![image](https://user-images.githubusercontent.com/1438610/227668022-3ec8358f-27c2-4500-b5a6-159a6a154b8f.png)

```
Julia Version 1.9.0-rc1
Commit 3b2e0d8fbc1 (2023-03-07 07:51 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, icelake-server)
  Threads: 8 on 8 virtual cores
Environment:
  JULIA_PKG_USE_CLI_GIT = true
  JULIA_NUM_THREADS = auto
  JULIA_EDITOR = code
```

**Comment by KristofferC** [comment]
Probably gonna need some code as well.

**Comment by BioTurboNick** [comment]
Found a minimal example. It appears to be when the length of an array is passed into a type parameter.

```
using BenchmarkTools
using StaticArrays
const xx = [1; 2]
@noinline function e(x)
    k = length(x)
    SVector{k}(x)
end

@btime e(xx) 
# 1.9-rc1: 27.898 μs (31 allocations: 2.66 KiB)
# 1.8.5: 762.239 ns (10 allocations: 544 bytes)
```

**Comment by BioTurboNick** [comment]
Linking my comment here: https://github.com/JuliaLang/julia/issues/48612#issuecomment-1484974814

TL;DR: If a performance regression is determined to be unavoidable for a particular release, can it be well-documented, users warned if affected pattern found, and/or could a workaround; or is some hacky patch possible to improve specific situations?

An aim being to reduce surprise and effort in troubleshooting if and when users hit it, and enable decision making when it comes to upgrading from 1.8 to 1.9.

**Comment by N5N3** [comment]
An easy fix on `StaticArrays.jl`'s side is replacing 
```julia
@propagate_inbounds (::Type{SA})(a::AbstractArray) where {SA <: StaticArray} = convert(SA, a)
```
with
```julia
@propagate_inbounds (T::Type{SA})(a::AbstractArray) where {SA <: StaticArray} = convert(T, a)
```
This make sure we can skip the unneeded runtime subtyping.

**Comment by vtjnash** [comment]
Yeah, we may need a heuristics fix for https://github.com/JuliaLang/julia/pull/45062, since right now it is likely pretty far off in the cost model

**Comment by vtjnash** [comment]
Appears fixed on master, although it may just be fixed in StaticArrays (https://github.com/JuliaArrays/StaticArrays.jl/pull/1157)
```
julia> @btime e(xx)
  897.279 ns (10 allocations: 480 bytes)
```
